### ISSUE TITLE
Kubernetes probes report failures when doing deployments

### ISSUE THREAD

**Issue by pecigonzalo** [sourcegraph/sourcegraph#13453]
When doing a deployment, some containers log `Pod indexed-search-4 in prod: Readiness probe failed: Get http://10.164.7.104:6070/healthz: read tcp 10.164.7.1:33362->10.164.7.104:6070: read: connection reset by peer`. 
Initially, we thought it was the initial readiness probes, which we increased the limit for https://github.com/sourcegraph/deploy-sourcegraph-dot-com/pull/3289 as our nodes take around 60s to start serving traffic which reduced the number of failure events.

While investigating further, we noticed the logs for `probe failed` come before the `Started container` logs and found a bug in Kubernetes which matches the behavior we are experiencing.

https://github.com/kubernetes/kubernetes/issues/52817

**Comment by slimsag** [comment]
@keegancsmith mentioned it might be possible to have Zoekt report it is ready during shard loading, could that help or do I misunderstand the issue here?

**Comment by pecigonzalo** [comment]
Its the opposite, the container is already shutting down, so its probably not handling requests anymore, but K8s still sends a `readinessProbe` check to that same container, which fails and we get a message in our Slack channel.

Aside from the value of `readinessProbe` conversation we had during our sync, I wonder if we could switch this to metric based alert which would allow to say `if X over a period of Y`

**Comment by slimsag** [comment]
OK, I see.

Looking closer after seeing your GitHub issue comment, I think this is the problem: the k8s event log isn't strictly errors, it also includes cases where readyiness probes rightfully failed because the pod is shutting down and the pod is being removed from the k8s watchers which we use to distribute load between the search-indexer replicas (note: there is actually "downed shard" logic here which begins to index the downed shard's dataset on the remaining shards that are alive automatically.)

Some thoughts:

1. Add an exception to https://github.com/sourcegraph/infrastructure/blob/master/docker-images/kubernetes-slack-bot/kubernetes-slack-bot.go#L74 which effectively silences this alert for search-indexer only.
2. Install a Prometheus metric there and consume that in our other alerting systems.

**Comment by pecigonzalo** [comment]
> it also includes cases where readyiness probes rightfully failed because the pod is shutting down
Yeah, the log is a Warning and not an Error, but I believe the expected behavior, as partially described by the upstream bug, is that the pod that is shutting down is removed from the replica list and `readinessProbe` stops checking as is known to be shutting down and its will most likely fail.

---

I would prefer the metric based option, as it would give us more flexibility on when to alert and streamline it with the rest of our observability stack. 

I will check if the following:
- Is it possible to correlate the failure with the pod being in "stopping" state?
- Is the pod effectively removed as soon as it start shutdown? (I believe this is the case, but I would like to confirm this to avoid a side effect)

If that probes complex, I will try to do the same correlation in the bot alert following the same logic. 
`pod.probe == failed && pod.status != stopping -> notify`

**Comment by keegancsmith** [comment]
What an annoying bug in kubernetes. If you read "terminating pods" section of kubernetes, when a pod is terminated it is removed from the endpoints list. IE readiness probe is an independent state that shouldn't matter anymore. From reading the bug report it does indeed seem like a bug rather than expected behaviour. But likely hard for k8s to fix given how it works.

I'd like to be able to rely on this signal since it has been very useful in the past. In general the kubernetes events have been useful for showing problems that occur which we didn't have monitoring for.

We could make our kubernetes slack bot ignore these events. We already do somewhat complicated logic around terminating states as well to prevent spamming on the bot starting up. So it has the information to do this since it receives the stream of events, so can ignore ready probe events for a terminating pod.

What is weird here is our `zoekt-webserver` doesn't listen for signals and then do a clean shutdown. It will just immediately exit upon receiving the signal. Maybe adding a clean shutdown would fix this? More likely not. My guess is we immediately exit but then the kernel needs to come in and clean up the thousands of open filehandles to all the mmap files which is why it isn't fast for the container to stop.

Alternatively maybe now that k8s has startup probes we can remove the readiness probe and instead just rely on startup + liveness probes. IE there should be no case where a `zoekt-webserver` should go away and then come back. Not sure if startup probes affect the inclusion in endpoints?

**Comment by pecigonzalo** [comment]
Indeed really annoying! My thought was to make the bot or metric ignore the alert if the pod is already in a terminating state, I believe this is the case in the public kuberenetes-mixin repo.

`startupProbe` :O I was not aware of this! I believe it could be a good switch for `zoekt`.

**Comment by keegancsmith** [comment]
So I just tested the change I made and we are still getting spam. I tailed the logs of a pod just before it got restarted and it was really interesting. It was receiving regular traffic (as witnessed by access logs), as soon as it received the SIGTERM from k8s the traffic pretty much disappeared. It waited 15s then shut itself down. 1s after shutting itself down the ready probe failing log message appeared in slack.

```
2020/10/16 19:50:54 received SIGTERM, waiting 15s before shutting down
2020/10/16 19:51:09 shutting down
```

<img width="481" alt="image" src="https://user-images.githubusercontent.com/187831/96303985-a2f36480-0ffb-11eb-8d6f-55f2afb5dd14.png">

This feels like a bug in k8s. Maybe this is related to the very high thresholds we give the ready check for zoekt? Surely more people would complain loudly if this was so easy to trigger? I suppose most people are not monitoring k8s events?

**Comment by pecigonzalo** [comment]
Yeah, that is be related to the linked upstream https://github.com/kubernetes/kubernetes/issues/52817 as I understand.

I think its because we are directly messaging on the events, my idea would be to try and replace those with metric based ones from Prometheus, hopefully we can match them there to pods not in Terminating state and avoid those failure messages.

**Comment by keegancsmith** [comment]
Yeah it does seem exactly related. Filtering based on state in our bot seems appropriate and doable.
